### ISSUE TITLE
ADB German Keyboard layout ^ and < reversed (Windows)

### ISSUE THREAD

**Issue by DOCaCola** [tmk/tmk_keyboard#35]
This is more or less a follow up on the older bug #22 that was fixed in the adb_usb converter.
I have now noticed something else. While these keys work as expected on my mac, They are reversed when the keyboard is plugged in on Windows (M3501 Extended II keyboard).
So when i press the ^ key (left of the 1 key), i get <.
If i press <, i get ^.
I thought it to be some kind of issue with Windows, so i googled and found this site (it is german though), but it might be helpful in analyzing this issue.
He describes some low level patch for the keyboard driver of Mac OS X.
http://snark.de/mac/usbkbpatch/
As it is already in the adb_usb code, he describes the adb keyboard codes as
0x32    <
0x0a    ^

but he has an interesting note on the usb codes for these keys:
0x64    ^ (caret - on wintel-usb switched with < (smaller) )
0x35    < (smaller - on wintel-usb switched with ^ (caret) )

Not sure what this means though. But there seems some kind of difference how Mac OS X interpretes these keys.

Switching the ADB code in the keymap.c obviously did the trick, but i guess a better solution exists, since other keyboards do not have this issue.

**Comment by tmk** [comment]
I found this post, seems iMate also has similar or worse problem with ISO keyboards.
https://geekhack.org/index.php?topic=4185.msg61270#msg61270

**Comment by tmk** [comment]
ISO AEK layout:
https://deskthority.net/w/images/1/18/AEKII-front.jpg
http://maximac.se/wordpress/wp-content/uploads/2014/12/apple-extended-keyboard-ii.jpg

US AEK layout:
https://deskthority.net/w/images/1/1a/Apple_M0115_top.jpg

**Comment by tmk** [comment]
how to determine keyboard layout with handler id.
http://lxr.free-electrons.com/source/drivers/macintosh/adbhid.c?v=4.4#L802

**Comment by tmk** [comment]
Added auto detection of ISO layout and swap of the problematic scan codes.
https://github.com/tmk/tmk_keyboard/commit/f43a2df7614503708818c258439d3a7c90d65e95

I hope this works for ISO keyboards.

**Comment by tmk** [comment]
As for this ISO keyboard auto detection test and report from ISO users are required.

**Comment by scytho** [comment]
I don’t know if this is still an issue.

I have an AEKII ISO German and use Pins 4 and 5 of a Teensy 2.0 for Data and Power. I didn’t change anything in the tmk code, so it should expect Data and Power on Pins 0 and 1. It works almost perfect with Linux Mint 17.3. Only »problem«: The equal sign on the Numpad generates an accent (»´«). I’d prefer a dot (».«, as alternative to the comma). < and ^ work as expected.

Same with Windows 8.1.

**Comment by scytho** [comment]
I changed the code to Pins 4 and 5, now < and ^ are switched. I changed it back to 0 and 1, but the Teensy only would go into bootloops and make the mouse pointer jump. After several tries and with Pins 4 and 5 in config.h it’s working again, < and ^ not switched. Weird.

**Comment by scytho** [comment]
After another programming of the Teensy < and ^ were switched again, plugging off and on the AEKII on the Teensy switched them back.

**Comment by kuerb** [comment]
I too have this issue with my German AEKII: NUBS and GRV are switched after programming the micro-controller. Once the keyboard is unplugged they are switched again, to whatever is defined in the keymap.

It seems like the auto detection only works if the keyboard is being programmed.

**Comment by tmk** [comment]
@kuerb yes, you are right. It detects ISO keyboard only once.

Fix and improvement for this is welcomed.

**Comment by egla89** [comment]
I have the same problem as a previous commentator where the EQUAL key on the numpad, registers as ‘ on an AEK2 iso. Mapped it as iso

**Comment by tmk** [comment]
@egla89 did you try firmware got from keyboard editor?
what is your keyboard layout on your os?

**Comment by egla89** [comment]
I flashed a teensy in my iso-de aek2 with ‘make -f Makefile.teensy KEYMAP=iso‘. Os is macOS 10.13 and I ran the layout recognition (so the < and ^ are in the right spot). Every other key works correctly

**Comment by tmk** [comment]
OK, then, key next to backspace(=) spits out equal correctly?

**Comment by tmk** [comment]
And is this pic what you referred as 'iso-de aek2'?
https://deskthority.net/w/images/1/18/AEKII-front.jpg

if not can you post pic of your aek2?

**Comment by kuerb** [comment]
![de_iso_aekii](https://user-images.githubusercontent.com/6669125/32100612-b973cdb0-bb14-11e7-86de-0794e77eec66.png)

Here is a picture of my German AEKII, which should be the same as the one egla89 is using.

In my own keymap I switched the numpads = key to PEQL. Maybe that's a simple fix for the general iso keymap?!

**Comment by tmk** [comment]
thanks for the pic and it makes sense.
`EQL` is recognized as `'` on german layout and numpad = key is defined as `EQL`in keymap_iso.c. `PEQL` may be less confused, in fact unimap_plain.c uses it.

anyway, you can change keymap as you want, there is no universal keymap for everyone.

**Comment by egla89** [comment]
Ok yes that makes sense. I‘ll change my layout accordingly. Thanks a lot!

**Comment by tmk** [comment]
Just for future reference, `PEQL` is not recognized on Windows and which is reason why some keymap use `EQL` for numpad '=' instead of `PEQL`.
https://github.com/tmk/tmk_keyboard/issues/238

**Comment by tmk** [comment]
Open new issue if you have problem.
